File: Java/백준/Gold/1987. 알파벳/알파벳.java
import java.io.*;
import java.util.*;

public class Main {
    static int R, C;
    static char[][] board;
    static boolean[] visited;
    static int max = 0;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        board = new char[R][C];
        for (int i = 0; i < R; i++) {
            String line = br.readLine();
            for (int j = 0; j < C; j++) {
                board[i][j] = line.charAt(j);
            }
        }

        visited = new boolean[26];
        dfs(0, 0, 1);
        System.out.println(max);
    }

    static void dfs(int x, int y, int count) {
        max = Math.max(max, count);
        visited[board[x][y] - 'A'] = true;

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx >= 0 && ny >= 0 && nx < R && ny < C) {
                if (!visited[board[nx][ny] - 'A']) {
                    dfs(nx, ny, count + 1);
                }
            }
        }

        visited[board[x][y] - 'A'] = false;
    }
}

File: C11/백준/Gold/1987. 알파벳/알파벳.c
#include <stdio.h>
#include <stdbool.h>

int R, C;
char board[20][21]; 
bool visited[26];
int max = 0;
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

void dfs(int x, int y, int count) {
    if (count > max) max = count;
    visited[board[x][y] - 'A'] = true;

    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];

        if (nx >= 0 && ny >= 0 && nx < R && ny < C) {
            if (!visited[board[nx][ny] - 'A']) {
                dfs(nx, ny, count + 1);
            }
        }
    }

    visited[board[x][y] - 'A'] = false;
}

int main() {
    scanf("%d %d", &R, &C);
    for (int i = 0; i < R; i++) {
        scanf("%s", board[i]);
    }

    dfs(0, 0, 1);
    printf("%d\n", max);

    return 0;
}